Add Darc dependency entries for our dependencies on Roslyn

jkoritzinsky opened the pull request:
Contributes to #65317 by making the main branch's Roslyn compiler dependency stay updated through darc. This way we don't need to keep remembering to update it.

cc: @stephentoub 

cc: @jaredpar as I'm effectively enshrining that dotnet/runtime will continue to use nightly Roslyn versions through the toolset package. We've been doing so since February and haven't seen any issues, but I was recommended to tag you in anyway to take a look.

jkoritzinsky committed:
Add Darc dependency entries for our dependencies on Roslyn

jkoritzinsky committed:
Enforce extended analyzer rules for generator projects in dotnet/runtime.

jkoritzinsky committed:
Update to latest build before opening the PR.

jkoritzinsky commented:
No more impact than our current usage of the `Microsoft.Net.Compilers.Toolset` package already has.

stephentoub commented:
Are you sure? I thought our current versioning of the compiler manually has been entirely separate from the versioning for the analyzer and style assemblies.

jkoritzinsky committed:
Organize dependencies by repo

stephentoub commented:
I'm fairly confident that our aggressive updating of MicrosoftNetCompilersToolsetVersion has not impacted required version in VS, whereas updating versions like MicrosoftCodeAnalysisCSharpCodeStyleVersion does.

jkoritzinsky commented:
> I'm fairly confident that our aggressive updating of MicrosoftNetCompilersToolsetVersion has not impacted required VS version, whereas updating versions like MicrosoftCodeAnalysisCSharpCodeStyleVersion has.

That would complicate this change, that's for sure.

Given how Roslyn uses darc and has a separate channel for each VS minor release, I see a few options here:

1. We keep the darc subscription always pointing at the most recently publicly released VS preview version channel, and manually update the toolset package to a newer version/a version in a different channel if need be. I don't think the darc tooling will fight us here, but I don't know for sure.
2. We have the toolset package update with a separate subscription that we can move faster. I don't think darc has the ability to do a split subscription like this though where we have enforced incoherency.

I'm also not sure how this will all interact with the VMR and source-build. IIRC, we disable analyzers in source-build due to this sort of problem, so we only need to keep the toolset up to date. With the VMR I guess we can depend on packages, so we don't need to use a live version from Roslyn.

jkoritzinsky commented:
Alternatively, I can drop the CodeStyle dependency from the DARC update and leave that one as a manually-updated dependency. The rest of the dependencies don't influence what runs in CI or only include analyzers that target downlevel Roslyn versions (dotnet/roslyn-analyzers targets pretty far down-level). @stephentoub what do you think of this idea?

stephentoub commented:
> @stephentoub what do you think of this idea?

I'm fine with it if it works.

jkoritzinsky committed:
Don't bump the code-style package version as raising that version raises our minimum VS version.

jkoritzinsky committed:
Update to most recent roslyn artifacts

ViktorHofer commented on eng/Version.Details.xml:
```diff
@@ -270,25 +270,21 @@
+      <Sha>91902d4dd2c07549fb64357ed5c3a0ee9d7c3610</Sha>
     </Dependency>
-    <Dependency Name="Microsoft.CodeAnalysis" Version="4.4.0-2.22463.8">
+    <Dependency Name="Microsoft.CodeAnalysis" Version="4.4.0-2.22464.20">
```
IIRC from my last conversation with @sharwell, bumping `Microsoft.CodeAnalysis` results in a newer VS version being required.

jkoritzinsky replied:
It depends where we use this reference. I need to check which package specifically we reference with the EventSourceGenerator as that’s the only generator that might reference this package and that we care about VS experience (though since it doesn’t affect the design time experience, having it silently fail in VS but still work in the build isn’t that bad). IIRC, we don’t care about the VS experience for the src/tests test projects, and the generator there doesn’t affect the design time experience anyway.

ViktorHofer replied:
Presumably a dev could add a reference to this dependency elsewhere in this repo and suddenly (without knowing) locks out devs from contributing that don't have access to the IntPreview VS builds. A good example would be in System.Private.CoreLib which ascendingly is referenced by many libraries and is part of many solution files under src/libraries. If today Microsoft.CodeAnalysis is only used in a component that wouldn't impact external contributors then the dependency is probably OK to have. At the least we need comments in Versions.props and maybe also in a markdown file explaining this dependency and what to watch out for.

ViktorHofer commented on src/libraries/Directory.Build.props:
```diff
@@ -52,11 +52,13 @@
     <AutoGenerateAssemblyVersion>true</AutoGenerateAssemblyVersion>
     <!-- To suppress warnings about resetting the assembly version.-->
     <AssemblyVersion />
+    <!-- Enforce extended rules around API usages in analyzers and generators to ensure our generators follow best practices. -->
+    <EnforceExtendedAnalyzerRules>true</EnforceExtendedAnalyzerRules>
```
For other reviewing this, here's the "documentation" that attempts to explain this setting. I don't see this officially anywhere on docs.microsoft.com though: https://github.com/dotnet/roslyn-analyzers/blob/03344a064a80c1f66e50b67e3abe7335cb8ba072/src/Microsoft.CodeAnalysis.Analyzers/Microsoft.CodeAnalysis.Analyzers.md#rs1036-specify-analyzer-banned-api-enforcement-setting

jkoritzinsky committed:
Update dependencies

jkoritzinsky committed:
Merge branch 'main' of github.com:dotnet/runtime into darc-dependency-roslyn

jkoritzinsky committed:
Add another "custom" manual version for MicrosoftCodeAnalysisVersion that is documented to specifically track the latest preview release of VS and use that version for any components that need to update similarly.

jkoritzinsky commented:
@ViktorHofer @stephentoub I've slightly changed my approach in this PR and I think you two will like it. I've introduced a new "pinned" version for Microsoft.CodeAnalysis named `MicrosoftCodeAnalysisVersion_LatestVS` that is documented that the infra team will not move forward of the version in the most recent public VS preview release. Any components that will be loaded into VS should use this property or an earlier version (and I've updated the two components and another package that don't but should). This version will automatically switch to the "live" version in source-build to ensure that we have a good source-build experience (and that our CI will catch more "live Roslyn" issues earlier).

The remaining versions that will be auto-updated will not cause any degradation to the local dev experience in VS.

jkoritzinsky committed:
Update to today's versions

jkoritzinsky committed:
Merge remote-tracking branch 'dotnet/main' into darc-dependency-roslyn
